File: src/freelm/_state.py
"""Opt-in persistent key state (rpd counters, cooldowns, disabled flags).

Survives process restarts so a fresh run doesn't re-burn keys that are already
exhausted or dead. JSON file in the cache dir, 0600, atomic replace. The schema
is shared with the JS package (``js/src/state.ts``) so both can read it:

    {"<provider>:<sha256(key)[:12]>": {"rpd_used": int, "rpd_reset_wall": float,
     "cooldown_until_wall": float, "disabled": bool, "last_error": str|null}}

Raw keys are never written — only a short hash. Wall-clock timestamps in the
file are converted to/from the in-process monotonic clock on load/save.
Multi-process use is last-writer-wins (best effort, documented).
"""
from __future__ import annotations

import hashlib
import json
import os
import time
from typing import Any, Dict, List, Optional

from ._cache import cache_dir


def _key_id(provider_name: str, key: str) -> str:
    return f"{provider_name}:{hashlib.sha256(key.encode('utf-8')).hexdigest()[:12]}"


class StateStore:
    def __init__(self, path: Optional[str] = None) -> None:
        self.path = path or os.path.join(cache_dir(), "state.json")

    def _read(self) -> Dict[str, Any]:
        try:
            with open(self.path, "r", encoding="utf-8") as f:
                data = json.load(f)
            return data if isinstance(data, dict) else {}
        except (OSError, ValueError):
            return {}

    def load_into(self, providers: List[Any], now_mono: float) -> None:
        data = self._read()
        if not data:
            return
        wall = time.time()
        for p in providers:
            for k in p.keys:
                e = data.get(_key_id(p.name, k.key))
                if not isinstance(e, dict):
                    continue
                k.rpd_used = int(e.get("rpd_used") or 0)
                rr = float(e.get("rpd_reset_wall") or 0)
                if rr > wall:
                    k.rpd_reset = now_mono + (rr - wall)
                cu = float(e.get("cooldown_until_wall") or 0)
                if cu > wall:
                    k.cooldown_until = now_mono + (cu - wall)
                k.disabled = bool(e.get("disabled", False))
                if e.get("last_error"):
                    k.last_error = e["last_error"]

    def save(self, providers: List[Any], now_mono: float) -> None:
        # merge over existing entries so other processes/providers aren't clobbered
        data = self._read()
        wall = time.time()
        for p in providers:
            for k in p.keys:
                data[_key_id(p.name, k.key)] = {
                    "rpd_used": k.rpd_used,
                    "rpd_reset_wall": wall + (k.rpd_reset - now_mono) if k.rpd_reset > 0 else 0,
                    "cooldown_until_wall": wall + (k.cooldown_until - now_mono) if k.cooldown_until > now_mono else 0,
                    "disabled": k.disabled,
                    "last_error": k.last_error,
                }
        tmp = self.path + ".tmp"
        try:
            os.makedirs(os.path.dirname(self.path), exist_ok=True)
            with open(tmp, "w", encoding="utf-8") as f:
                json.dump(data, f)
            os.chmod(tmp, 0o600)
            os.replace(tmp, self.path)
        except OSError:
            pass  # persistence is best-effort; never fatal

File: js/src/state.ts
/** Opt-in persistent key state (rpd counters, cooldowns, disabled flags).
 *
 * Survives process restarts so a fresh run doesn't re-burn keys that are
 * already exhausted or dead. JSON file in the cache dir, 0600, atomic replace.
 * The schema is shared with the Python package (`src/freelm/_state.py`):
 *
 *   {"<provider>:<sha256(key)[:12]>": {rpd_used, rpd_reset_wall,
 *    cooldown_until_wall, disabled, last_error}}
 *
 * Raw keys are never written — only a short hash. Wall-clock timestamps in the
 * file are converted to/from the in-process monotonic clock on load/save.
 * Multi-process use is last-writer-wins (best effort, documented).
 */
import { createHash } from "node:crypto";
import * as fs from "node:fs";
import * as path from "node:path";
import { cacheDir } from "./cache.js";
import { wallS } from "./time.js";

function keyId(providerName: string, key: string): string {
  return `${providerName}:${createHash("sha256").update(key, "utf-8").digest("hex").slice(0, 12)}`;
}

export class StateStore {
  path: string;

  constructor(filePath?: string) {
    this.path = filePath ?? path.join(cacheDir(), "state.json");
  }

  private read(): Record<string, any> {
    try {
      const data = JSON.parse(fs.readFileSync(this.path, "utf-8"));
      return data && typeof data === "object" && !Array.isArray(data) ? data : {};
    } catch {
      return {};
    }
  }

  loadInto(providers: any[], nowMono: number): void {
    const data = this.read();
    if (!Object.keys(data).length) return;
    const wall = wallS();
    for (const p of providers) {
      for (const k of p.keys) {
        const e = data[keyId(p.name, k.key)];
        if (!e || typeof e !== "object") continue;
        k.rpdUsed = Math.trunc(Number(e.rpd_used) || 0);
        const rr = Number(e.rpd_reset_wall) || 0;
        if (rr > wall) k.rpdReset = nowMono + (rr - wall);
        const cu = Number(e.cooldown_until_wall) || 0;
        if (cu > wall) k.cooldownUntil = nowMono + (cu - wall);
        k.disabled = Boolean(e.disabled);
        if (e.last_error) k.lastError = String(e.last_error);
      }
    }
  }

  save(providers: any[], nowMono: number): void {
    // merge over existing entries so other processes/providers aren't clobbered
    const data = this.read();
    const wall = wallS();
    for (const p of providers) {
      for (const k of p.keys) {
        data[keyId(p.name, k.key)] = {
          rpd_used: k.rpdUsed,
          rpd_reset_wall: k.rpdReset > 0 ? wall + (k.rpdReset - nowMono) : 0,
          cooldown_until_wall: k.cooldownUntil > nowMono ? wall + (k.cooldownUntil - nowMono) : 0,
          disabled: k.disabled,
          last_error: k.lastError,
        };
      }
    }
    const tmp = this.path + ".tmp";
    try {
      fs.mkdirSync(path.dirname(this.path), { recursive: true });
      fs.writeFileSync(tmp, JSON.stringify(data), { mode: 0o600 });
      fs.renameSync(tmp, this.path);
    } catch {
      // persistence is best-effort; never fatal
    }
  }
}
